### ISSUE TITLE
Sequence(start, stop, [step])

### ISSUE THREAD

**Issue by leandromoh** [morelinq/MoreLINQ#240]
`Enumerable.Range` only can generate integer sequences that are ascending and with step +1 between the elements. Anything different becomes incredibly painful.

For example, for generate the sequence `[10000, 9999, ..., -10000]` you must write: 
`Enumerable.Range(-10000, 20001).Reverse()`

If your sequence has step different of 1 or -1, you simply can't generate it.

My propose is to add the following overloads to this function:

- `IEnumerable<int> Range(int start, int step)`
- `IEnumerable<int> Range(int start, int step, int limit)`

So you can create more flexibles integer sequences much simpler. The previous example can be rewrite with:

`Range(10000, -1).Take(20001)` or even `Range(10000, -1, -10000)`.

I'm planning to submit a PR for this.

What do you think?

**Comment by atifaziz** [comment]
There's already [`Generate`](http://morelinq.github.io/1.x/ref/api/html/M_MoreLinq_MoreEnumerable_Generate__1.htm) for that and should cover all the cases you pointed out:

So for `[10000, 9999, ..., -10000]`, you can do this:

```c#
MoreEnumerable.Generate(10000, i => i - 1).TakeWhile(i => i >= -10000)
```

> If your sequence has step different of 1 or -1, you simply can't generate it.

With `Generate`, you can:

```c#
MoreEnumerable.Generate(10, i => i - 2).TakeWhile(i => i >= -10)
```

Needless to say, since `Generate` is generic, it can also work on non-integers:

```c#
// Every half hour of the last 4 days
MoreEnumerable.Generate(DateTime.Today.AddDays(-4), d => d.AddHours(0.5))
              .TakeWhile(d => d < DateTime.Today)
```

**Comment by leandromoh** [comment]
> There's already `Generate `for that and should cover all the cases you pointed out.

I know, but i think that 

`Range(10000, -1, -10000)`

Is much more readable and clear than

`MoreEnumerable.Generate(10000, i => i - 1).TakeWhile(i => i >= -10000)`.

Other disadvantage of the above approach is the unnecessary complexity, O(n) + O(n) I guess, since for every element in the result `IEnumerable`, we had to iterate it twice (one in `Generate` and other in `TakeWhile`).

> since Generate is generic, it can also work on non-integers:

Really, this is a greate benefit.

The others overloads really isn't necessary, would be just for readability and reduce effort.

**Comment by leandromoh** [comment]
Thinking how to solve the problem of unnecessary complexity and keeping generics, I remember about the `unfold` function (present and many functional like Haskell and F#), that is a dual to `fold` and very similar to `Generate` (`Generate` is a specification of `unfold` for infinit lists), but letting the function that it receives to decide when the resulting `IEnumerable` ends. This resolve the problem of unnecessary complexity and keep generic spirit alive.

`unfold` signature in Haskell and F# respectively:

- `unfoldr :: (b -> Maybe (a, b)) -> b -> [a]`
- `Seq.unfold : ('State -> ('T * 'State) option) -> 'State -> seq<'T>`

since C# dont have `Maybe/option` type we could expect a triple instead of tuple, being the last element a `bool `indicating if must continue the enumeration or not. other approanch is separete the triple in 3 functions. Something like:

- `IEnumerable<TResult> UnFold<TSource, TResult>(TSource seed, Func<TSource, Tuple<TResult, TSource, bool>> generator)`
- `IEnumerable<TResult> UnFold<TSource, TResult>(TSource seed, Func<TSource, bool> predicate, Func<TSource, TResult> generateResult, Func<TSource, TSource> generateSeed)`

@atifaziz What do you think about add `unfold`  (maybe as `UnAggregate`) on MoreLinq?

**Comment by atifaziz** [comment]
If you want to add a simpler version, I'd say better to call it `Sequence` instead of `Range`. It would help to avoid confusion with [`Enumerable.Range`](https://msdn.microsoft.com/en-us/library/system.linq.enumerable.range(v=vs.110).aspx). I imagine the signatures as follows:

```c#
public static IEnumerable<int> Sequence(int start, int stop) { /* ... */ }
public static IEnumerable<int> Sequence(int start, int stop, int step) { /* ... */ }
```

It would also make sense (in the spirit of simplicity) for `Sequence` to automatically negate the step in the first overload if `start` is greater than `stop`, such that `Sequence(4, 1)` yields `[4,3,2,1]`. It should not do this if `step` is explicitly supplied via the second overload and should yield nothing if the step direction is wrong.

If that sounds right, feel free to submit a PR.

As for unfolding, let's take that up as a separate issue instead of overloading the discussion here.

**Comment by leandromoh** [comment]
`If that sounds right, feel free to submit a PR.`

Ok, I will submit.

**Comment by atifaziz** [comment]
Cool. Should we aim to get this into 2.2 that I've slated for end of month?

**Comment by leandromoh** [comment]
> Should we aim to get this into 2.2 that I've slated for end of month?

Yes, i can do it until the end of month. 
How about `AggregateRight `and `SkipLast`? They're ready finished.

**Comment by atifaziz** [comment]
[`SkipLast` is currently slated for 2.2](https://github.com/morelinq/MoreLINQ/milestone/3). I'll take a look at `AggregateRight`.

**Comment by atifaziz** [comment]
Now [published as 2.2 beta 1](https://www.nuget.org/packages/morelinq/2.2.0-beta-1).

**Comment by atifaziz** [comment]
[Documentation is also published](https://morelinq.github.io/2.2/ref/api/html/Overload_MoreLinq_MoreEnumerable_Sequence.htm).
